app/estimate_square_size.go

```diff
@@ -0,0 +1,246 @@
+			// invalid (cannot be wrapped)
+			if err != nil {
+				panic(err)
+			}
+			used, _ := shares.MsgSharesUsedNIDefaults(msgSharesCursor, squareSize, tx.msg.Size())
+			msgSharesCursor += used
+		}
+		txSplitter.WriteTx(rawTx)
+	}
+	for _, e := range evd.Evidence {
+		evidence, err := coretypes.EvidenceFromProto(&e)
+		if err != nil {
+			panic(err)
+		}
+		err = evdSplitter.WriteEvidence(evidence)
+		if err != nil {
+			panic(err)
+		}
+	}
+	txCount, available := txSplitter.Count()
+	if consts.TxShareSize-available > 0 {
+		txCount++
+	}
+	evdCount, available := evdSplitter.Count()
+	if consts.TxShareSize-available > 0 {
+		evdCount++
+	}
+	return txCount + evdCount
+}
+
+// estimateSquareSize uses the provided block data to estimate the square size
+// assuming that all malleated txs follow the non interactive default rules.
+// Returns the estimated square size and the number of shares used.
+func estimateSquareSize(txs []*parsedTx, evd core.EvidenceList) (uint64, int) {
+	// get the raw count of shares taken by each type of block data
+	txShares, evdShares, msgLens := rawShareCount(txs, evd)
+	msgShares := 0
+	for _, msgLen := range msgLens {
+		msgShares += msgLen
+	}
+
+	// calculate the smallest possible square size that could contain all the
+	// messages
+	squareSize := nextPowerOfTwo(int(math.Ceil(math.Sqrt(float64(txShares + evdShares + msgShares)))))
+
+	// the starting square size should be the minimum
+	if squareSize < consts.MinSquareSize {
+		squareSize = consts.MinSquareSize
+	}
+
+	var fits bool
+	for {
+		// assume that all the msgs in the square use the non-interactive
+		// default rules and see if we can fit them in the smallest starting
+		// square size. We start the cusor (share index) at the begginning of
+		// the message shares (txShares+evdShares), because shares that do not
+		// follow the non-interactive defaults are simple to estimate.
+		fits, msgShares = shares.FitsInSquare(txShares+evdShares, squareSize, msgLens...)
+		switch {
+		// stop estimating if we know we can reach the max square size
+		case squareSize >= consts.MaxSquareSize:
+			return consts.MaxSquareSize, txShares + evdShares + msgShares
+		// return if we've found a square size that fits all of the txs
+		case fits:
+			return uint64(squareSize), txShares + evdShares + msgShares
+		// try the next largest square size if we can't fit all the txs
+		case !fits:
+			// increment the square size
+			squareSize = nextPowerOfTwo(squareSize + 1)
+		}
+	}
+}
+
+// rawShareCount calculates the number of shares taken by all of the included
+// txs, evidence, and each msg. msgLens is a slice of the number of shares used
+// by each message without accounting for the non-interactive message layout
+// rules.
+func rawShareCount(txs []*parsedTx, evd core.EvidenceList) (txShares, evdShares int, msgLens []int) {
+	// msgSummary is used to keep track of the size and the namespace so that we
+	// can sort the namespaces before returning.
```

Comment:
```suggestion
	// msgSummary is used to keep track of the size and the namespace so that we
	// can sort the messages by namespace before returning.
```